File: Problem3.py
# Time Complexity: O(m+n)
# Space Complexity: O(1)
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
      r, c = len(matrix) - 1, 0
      while r >= 0 and c <= len(matrix) - 1:
        if matrix[r][c] == target:
          return True
        elif matrix[r][c] > target: # larger than up a row
          r -= 1
        else: # smaller then move to the right
          c += 1
      return False
      
#BF: O(m*n)
      # for r in range(len(matrix)):
      #   for c in range(len(matrix[0])):
      #     if matrix[r][c] == target:
      #       return True
      # return False

      # O(m*logn)
      # Bsearch instead of loop
      # def bsearch(nums, target):
      #   l, r = 0, len(nums) - 1
      #   while l <= r:
      #     mid = l + ((r-l) // 2)
      #     m = nums[mid]
      #     if m == target:
      #       return True
      #     elif m < target:
      #       l = mid + 1
      #     else:
      #       r = mid - 1
      #   return False

      # for r in range(len(matrix)):
      #   exists = bsearch(matrix[r], target)
      #   if exists:
      #     return True

      # return False
      

      
      

File: Problem_3.js
// Problem3 (https://leetcode.com/problems/search-a-2d-matrix-ii/)

// Time Complexity : O(m+n) where m is rows and n is columns
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
// Start from top right element
// If target is less than number at location, move down as the answer will be larger
// Else move left

// Your code here along with comments explaining your approach

/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function (matrix, target) {
    if (matrix === null || matrix.length === 0) {
        return false;
    }
    // Start from top right element
    // If target is less than number at location, move down as the answer will be larger
    // Else move left
    let i = 0;
    let j = matrix[0].length - 1;
    while (i < matrix.length && j >= 0) {
        if (matrix[i][j] === target) {
            return true;
        } else if (matrix[i][j] < target) {
            // Go down
            i++;
        } else {
            // Go left
            j--;
        }
    }
    return false;
};